SCons/CacheDir.py

```diff
@@ -50,18 +51,28 @@ def CacheRetrieveFunc(target, source, env):
         cd.CacheDebug('CacheRetrieve(%s):  %s not in cache\n', t, cachefile)
         return 1
     cd.hits += 1
-    cd.CacheDebug('CacheRetrieve(%s):  retrieving from %s\n', t, cachefile)
     if SCons.Action.execute_actions:
         if fs.islink(cachefile):
             fs.symlink(fs.readlink(cachefile), t.get_internal_path())
         else:
-            env.copy_from_cache(cachefile, t.get_internal_path())
+            try:
+                env.copy_from_cache(cachefile, t.get_internal_path())
+            except (shutil.SameFileError, IOError) as e:
+                    # In case file was partially retrieved (and now corrupt)
+                    # delete it to avoid poisoning commands like 'ar' that
+                    # read from the initial state of the file they are writing
+                    # to.
+                    t.fs.unlink(t.get_internal_path())
+                    cd.CacheDebug('CacheRetrieve(%s):  Error while retrieving from %s deleting %s\n', t, cachefile)
+                    raise e
```

Comment:
Here shouldn't we behave as if the file were not in the cache, i.e. return 1, rather than raise?